File: sudoku.py
from PySide2.QtCore import QThread, Signal
import time

class Sudoku(QThread):
    signal_show_cell = Signal(object, object, object)
    signal_hide_cell = Signal(object, object)
    signal_solved = Signal()

    def __init__(self):
        super(Sudoku, self).__init__()
        
        self.matrix = list()
        self.DH = []
        self.delay = 0.05
        self.preview = False

    def run(self):
        self.degreeCalculation()
        if self.Sudoku():
            print("Done!")
            self.signal_solved.emit()
        else:
            print("No solution exists")

    def degreeCalculation(self):
        for i in range(9):
            for j in range(9):
                cntr = 0
                if self.matrix[i][j] == 0:
                    for k in range(9):
                        if (self.matrix[i][k] == 0 and k != j) or (self.matrix[k][j] == 0 and k != i):
                            cntr += 1

                    for k in range((i // 3) * 3, ((i // 3) * 3) + 3):
                        for m in range((j // 3) * 3, ((j // 3) * 3) + 3):
                            if self.matrix[k][m] == 0 and k != i and m != j:
                                cntr += 1
                    
                    node = {'degree':cntr, 'row':i, 'col':j, 'isset':0}
                    self.DH.append(node)

        self.DH = sorted(self.DH, key=lambda k: k['degree'])

    def usedInRow(self, row, num):
        if all(self.matrix[row][col] != num for col in range(9)):
            return False
        return True

    def usedInCol(self, col, num):
        if all(self.matrix[row][col] != num for row in range(9)):
            return False
        return True

    def usedInBox(self, boxStartrow, boxStartcol, num):
        if all(self.matrix[row + boxStartrow][col + boxStartcol] != num for col in range(3) for row in range(3)):
            return False
        return True

    def isSafe(self, row, col, num):
        return not self.usedInRow(row, num) and not self.usedInCol(col, num) and \
               not self.usedInBox(row - row % 3, col - col % 3, num)

    def Heuristic(self):
        for node in self.DH:
            if node['isset'] == 0:
                row = node['row']
                col = node['col']
                return True, row, col
        return False, 0, 0

    def Set(self, row, col):
        this = next(item for item in self.DH if row == item['row'] and col == item['col'])
        this['isset'] = 1

        for node in self.DH:
            if this['row'] == node['row'] or this['col'] == node['col'] or (
                    this['row'] // 3 == node['row'] // 3 and this['col'] // 3 == node['col'] // 3):
                node['degree'] -= 1

        self.DH = sorted(self.DH, key=lambda k: k['degree'])

    def UnSet(self, row, col):
        this = next(item for item in self.DH if row == item['row'] and col == item['col'])
        this['isset'] = 0

        for node in self.DH:
            if this['row'] == node['row'] or this['col'] == node['col'] or (
                    this['row'] // 3 == node['row'] // 3 and this['col'] // 3 == node['col'] // 3):
                node['degree'] += 1

        self.DH = sorted(self.DH, key=lambda k: k['degree'])

    def Sudoku(self):
        try:
            flag, row, col = self.Heuristic()

            if not flag:
                return True

            for num in range(1, 10):
                if self.isSafe(row, col, num):

                    self.Set(row, col)
                    self.matrix[row][col] = num
                    if self.preview:
                        time.sleep(self.delay)
                        self.signal_show_cell.emit(row, col, num)

                    if self.Sudoku():
                        return True

                    self.UnSet(row, col)
                    self.matrix[row][col] = 0
                    if self.preview:
                        time.sleep(self.delay)
                        self.signal_hide_cell.emit(row, col)

            return False

        except:
            print('some problem')

        return True

File: source_cpp/sudoku.cpp
#include "iostream"
#include "QtDebug"
#include <source_cpp/sudoku.h>


using namespace std;


Sudoku::Sudoku()
{
    delay = 10;
    preview = false;
}


Sudoku::~Sudoku()
{

}


void Sudoku::run()
{
    DH.clear();
    degreeCalculation();

    if (solve() == true)
    {
        qDebug() << "solved";
        emit solved();
    }
    else
        qDebug() << "No solution exists";
}


void Sudoku::sort()
{
    for (int i = 0; i < DH_size; i++)
    {
        for (int j = i + 1; j < DH_size; j++)
        {
            if (DH[i].degree > DH[j].degree)
                swap(DH[i], DH[j]);
        }
    }
}


void Sudoku::degreeCalculation()
{
    int cntr;
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            if (matrix[i][j] == UNASSIGNED)
            {
                cntr = 0;
                DegreeHeuristic dh;
                for (int k = 0; k < N; k++)
                {
                    if (matrix[i][k] == UNASSIGNED && k != j)
                    {
                        cntr++;
                    }
                    if (matrix[k][j] == UNASSIGNED && k != i)
                    {
                        cntr++;
                    }
                }
                for (int k = (i / 3) * 3; k < ((i / 3) * 3) + 3; k++)
                {
                    for (int m = (j / 3) * 3; m < ((j / 3) * 3) + 3; m++)
                    {
                        if (matrix[k][m] == UNASSIGNED && k != i && m != j)
                        {
                            cntr++;
                        }
                    }
                }
                dh.degree = cntr;
                dh.Row = i;
                dh.Col = j;
                dh.IsSet = 0;
                DH.push_back(dh);
            }
        }
    }
    DH_size = DH.size();
    sort();
}


bool Sudoku::heuristic(int &Row, int &Col)
{
    for (int i = 0; i < DH_size; i++)
    {
        if (DH[i].IsSet == 0)
        {
            Row = DH[i].Row;
            Col = DH[i].Col;
            return true;
        }
    }
    return false;
}


void Sudoku::set(int Row, int Col)
{
    for (int i = 0; i < DH_size; i++)
    {
        if (Row == DH[i].Row && Col == DH[i].Col)
        {
            DH[i].IsSet = 1;

            for (int j = 0; j < DH_size; j++)
            {
                if (DH[i].Row == DH[j].Row || DH[i].Col == DH[j].Col || (DH[i].Row / 3 == DH[j].Row / 3 && DH[i].Col / 3 == DH[j].Col / 3))
                {
                    DH[j].degree--;
                }
            }
            sort();
            break;
        }
    }
}


void Sudoku::unSet(int Row, int Col)
{
    for (int i = 0; i < DH_size; i++)
    {
        if (Row == DH[i].Row && Col == DH[i].Col)
        {
            DH[i].IsSet = 0;

            for (int j = 0; j < DH_size; j++)
            {
                if (DH[i].Row == DH[j].Row || DH[i].Col == DH[j].Col || (DH[i].Row / 3 == DH[j].Row / 3 && DH[i].Col / 3 == DH[j].Col / 3))
                {
                    DH[j].degree++;
                }
            }
            sort();
            break;
        }
    }
}


void Sudoku::setPreview(bool b)
{
    preview = !preview;
}


bool Sudoku::solve()
{
    int row, col;
    if (!heuristic(row, col))
        return true;

    for (int num = 1; num <= N; num++)
    {
        if (isSafe(row, col, num))
        {
            set(row, col);
            matrix[row][col] = num;

            if(preview)
            {
                msleep(delay);
                emit signalShowCell(row, col, num);
            }

            if (solve())
                return true;

            unSet(row, col);
            matrix[row][col] = UNASSIGNED;

            if(preview)
            {
                msleep(delay);
                emit signalHideCell(row, col);
            }
        }
    }
    return false;
}


bool Sudoku::usedInRow(int Row, int num)
{
    for (int Col = 0; Col < N; Col++)
        if (matrix[Row][Col] == num)
            return false;
    return true;
}


bool Sudoku::usedInCol(int Col, int num)
{
    for (int Row = 0; Row < N; Row++)
        if (matrix[Row][Col] == num)
            return false;
    return true;
}


bool Sudoku::usedInBox(int boxStartRow, int boxStartCol, int num)
{
    for (int Row = 0; Row < 3; Row++)
        for (int Col = 0; Col < 3; Col++)
            if (matrix[Row + boxStartRow][Col + boxStartCol] == num)
                return false;
    return true;
}


bool Sudoku::isSafe(int Row, int Col, int num)
{
    return usedInRow(Row, num) && usedInCol(Col, num) && usedInBox(Row - Row % 3, Col - Col % 3, num);
}
